### DISCUSSION TITLE
Process to migrate data from MariaDB to PostgreSQL?

### DISCUSSION THREAD

**Discussion by ameyp** [firefly-iii/firefly-iii#7698] [Firefly III] [answered]
I've been running Firefly-III for about a year now with a separate MariaDB database. All other self-hosted services I use are on PostgreSQL. I ran into some infrastructure issues yesterday with the MariaDB server, and decided that I'd rather operate one database server than two. Is there an easy way to migrate all my data from the existing database to PostgreSQL? I tried using https://github.com/dimitri/pgloader/ but it reported 17 total `Create Foreign Keys` errors while transferring data, e.g.:

```
foreign key constraint "type_user_group_id" cannot be implemented
```

**Comment by JC5** [comment]
Mmm, I would have referred you to such a tool. Can you also export/import just the data? Let Firefly III regenerate its own database and then import only the data.

**Reply by ameyp** [reply]
How would I import the data? https://docs.firefly-iii.org/firefly-iii/exporting-data/ says 

> Firefly III does not have the functionality to import this data. If you need to back up or migrate your data, backup and restore your database directly.

Can https://docs.firefly-iii.org/data-importer/how-to-use/ handle the data exported from Firefly? Is there an order in which it would need to be imported? For example, tags before transactions or something like that.

**Reply by JC5** [reply]
No, it can't, that's why it has the big fat warning. What I'm asking is, could such a tool as you suggest import+export the data alone?

**Comment by Adrian-at-CrimsonAzure** [comment, marked as answer]
I have completed this process using [pgloader](https://pgloader.io/) and everything _appears_ to be working so far. Here's what I did:
1. Create a duplicate Firefly instance with an identical .env
2. Edit the env to point to a Postgres database
3. Start the new stack/container/chart/whatever and wait until you can get to the registration page
4. Create a schema-only backup using PGAdmin
![image](https://github.com/firefly-iii/firefly-iii/assets/1719461/ad4314b6-cb03-4a46-aa7f-6360ca7c44c8)
5. Restore the schema-only backup with the "clean before restore" option selected. As far as I can tell this empties the tables when you do the restore.
![image](https://github.com/firefly-iii/firefly-iii/assets/1719461/0feb35d8-16dc-442b-a388-4dbd4c9caf3d)
The reason the backup/restore step is needed is that there are a few records spread across the tables that are automatically created when Firefly first initializes. Whatever these records are, they were already created by your initial instance so they can be safely deleted and this is the easiest way I found to reset the database to an entirely empty state. 

6. Install pgloader on a system with access to both databases and create this `loader.txt` configuration file:
```
LOAD DATABASE
        FROM mysql://[FIREFLY_DB_USER]:[FIREFLY_DB_PASSWORD]@[FIREFLY_DB_IP_ADDRESS]/[FIREFLY_DB_NAME]
        INTO pgsql://[FIREFLY_DB_USER]:[FIREFLY_DB_PASSWORD]@[FIREFLY_DB_IP_ADDRESS]/[FIREFLY_DB_NAME]
        WITH data only, quote identifiers
        ALTER SCHEMA 'firefly' RENAME TO 'public'
;
```
`data only` will ensure the table schema is left untouched, and `quote identifiers` preserves the case-sensitive field names. By default pgloader will place your database into a schema of the same name but, like most applications, Firefly uses the `public` schema so the `ALTER` line is needed

7.  Run `pgloader loader.txt` and ignore any warnings. If you see a chart at the end then it was successful
8. Exec into the container and run `php artisan firefly-iii:correct-database`. I'm not sure if this is needed but it seemed like it did something and I wanted to be safe
9. Restart the Postgres-enabled Firefly instance and log in

If a few more users can verify this it would be nice to have this mentioned somewhere in the documentation somewhere as an unofficial migration path.

**Reply by JC5** [reply]
That's an amazing write-up, thanks so much! I'll be sure to put it in the documentation!

**Reply by elizhyu** [reply]
I just had a successful migration from `MariaDB` to `Postgres`.
I'm using docker compose for hosting my services, so here are some additional tips/steps/notes regarding setups like mine.

- Set all services in the same `network` and give services `hostname` like below: **(only worth-noticing lines are added)**

```
services:
  db:
    image: postgres
    hostname: fireflyiiidb
    ports:
      - 5432:5432
    networks:
      - firefly_iii
 
  db-old:
    image: mariadb:latest
    hostname: fireflyiiidbold
    ports:
      - 3306:3306
    networks:
      - firefly_iii
```
- Run `PGAdmin` through docker like this:
```
docker run -p 5050:80 --network firefly-iii_firefly_iii \
-e "PGADMIN_DEFAULT_EMAIL=user@domain.com" \
-e "PGADMIN_DEFAULT_PASSWORD=SuperSecret" \
-d dpage/pgadmin4
```
- Use `apt install` for the `pgloader` instead of `docker`
  _(I haven't figured out how to load the configuration file shown above into pgloader within docker, but if you succeeded, ignore what I'm saying here. Also if possibile, please kindly leave how below.)_
- Replace the `ALTER SCHEMA 'firefly' RENAME TO 'public'` line with your own database name, in my case: `ALTER SCHEMA 'firefly-iii-data' RENAME TO 'public'`
- Follow strictly the order like this: (Wrong order had failed me, and I had to start everything over)
  - Run an empty `Postgres` container to let it generate all the configuration/data files
  - Run the `firefly-iii` container using the above **empty** `Postgres` db, until you see the `firefly-iii` log outputs its log and `ready to use` message.
    (The failing point I've encountered here is that if not running the `firefly-iii` container, it will not create the `public` schema in `Postgres`, which will fail the `pgloader` script)
  - Run the `PGAdmin` container, and `backup` & `restore` as shown above.
    (The failing point I've encountered here is that if not cleaning everything while keeping the schema, all of my migrated transactions would use a weird currency other than my default currency USD, which would be such a pain considering there is currently no easy way to change the currency of all transactions. Maybe it's easy to do that with database operations, but I'm not an expert on that.)
  - Run `php artisan firefly-iii:correct-database` in the `firefly-iii` container.
    (It will be stuck at one point in my case, but it turns out being just slow. So be patient about this and wait until it goes all the way through.)
  - Restart the `firefly-iii` container, and it's all done.

**Reply by TjenWellens** [reply]
followed @Adrian-at-CrimsonAzure's steps to do a successful migration!

some useful snippets for people who started with the docker-compose setup:

! careful to **NEVER** do a `docker-compose down`, because this **deletes your volumes** with all your data. At most run `docker-compose stop/start`.

mariadb/.db.env  # no changes from https://docs.firefly-iii.org/how-to/firefly-iii/installation/docker/
```
MYSQL_RANDOM_ROOT_PASSWORD=yes
MYSQL_USER=firefly
MYSQL_PASSWORD=secret_firefly_password
MYSQL_DATABASE=firefly
```

After `docker-compose stop`, I ran a docker run to get access to the db from the compose stack (by mounting in the volume). 
```
docker run -it \
--env-file .db.env \
--env MYSQL_RANDOM_ROOT_PASSWORD=false \
-v firefly-iii-docker_firefly_iii_db:/var/lib/mysql \
-v /Users/tjen/dev/firefly/mariadb/db-bcp:/db-bcp \
-p 3306:3306 \
mariadb:lts
```
(I wanted to take a backup to a new adding volume on my laptop, requiring me to reset the root password with `--skip-grant-tables`, so running it outside compose seemed easiest)

postgres/.env
```
DB_CONNECTION=pgsql
DB_PORT=5432
```

postgres/.db.env
```
POSTGRES_USER=firefly
POSTGRES_PASSWORD=secret_firefly_password
POSTGRES_DB=firefly
```

postgres/docker-compose.yaml
```
  db:
    image: postgres
    hostname: db
    container_name: new_firefly_iii_db
    restart: always
    env_file: .db.env
    ports:
      - '5432:5432'
    networks:
      - firefly_iii
    volumes: 
      - firefly_iii_db:/var/lib/postgresql/data
      - /Users/tjen/dev/firefly/postgres/db-bcp:/db-bcp
```

[edited]
changed volume from `- firefly_iii_db:/var/lib/mysql` to `- firefly_iii_db:/var/lib/postgresql/data` to prevent future ppl's loss of data

**Reply by PAS-BC** [reply]
@TjenWellens "Amen' to never using 'docker compose down' - which of course I have recently done, so scrambling to get my backed up .csv data back.

However, from now on, I want to use a Share from my (separate) TrueNAS box as storage for FireflyIII running on my Proxmox box, but am struggling as a Linux NOOB.

My CIFS Share on TrueNAS is mounted via fstab and shows up in the Docker VM as '/mnt/truenas/folder-name-per-app' (it's finally working for Immich BTW).

If I substitute the '/mnt/...' for your '/Users/tjen/dev/firefly/postgres/db-bcp' above, would I be on the right track?

TIA

**Reply by TjenWellens** [reply]
Take what I say with a grain of salt, given that I don't really know (other than a vague idea) what these words mean: TrueNAS, Proxmox, CIFS, fstab. ;-)

Yes, substitute the '/mnt/...' for my '/Users/tjen/dev/firefly/postgres/db-bcp' should work.

1) You may want to bind the db volume `firefly_iii_db:/var/lib/mysql` to a path instead of a named volume. That way, a docker compose down, does not destroy the data. (eg. instead of `firefly_iii_db:/var/lib/mysql` I'd now replace it with `/Users/tjen/dev/firefly/postgres/data:/var/lib/mysql`, but see the next remark if you're using postgres)

2) my postgres/docker-compose.yaml volume (last snippet in my post) mounts the volume to `/var/lib/mysql`, I'm certain this is wrong, so edited it now. Not sure how that ended up there... Seems like a container restart (stop-start) could lose all it's data, since it's mounting the wrong dir for postgres. I likely got lucky, and did not stop the container, and never restarted my container. Since I only used it to create a backup file from the postgress db, and then loaded it into another one where I did mount the correct one: `/opt/fireflyiii/db-data:/var/lib/postgresql/data`

**Reply by PAS-BC** [reply]
Thank you for the feedback.

As a Windows refugee, I'm still trying to get my head around 'Bind, 'Mount' and 'Everything is a directory'.

I am however, now fully committed to FF and have ditched my YNAB Sub.
So I won't be rushing into migrating >5,000 transaction from MySQL to Postgres on a different machine in a hurry :)

**Reply by jpizquierdo** [reply]
thank you guys, it worked for me today :)

**Reply by c0l0nslashslash** [reply]
Works great, thanks for the guide

## Make sure to use pgloader version 3, version 4 will not work with this

**Comment by r0m1CH** [comment]
Hi, I'm looking for a way to migrate from Maria to Postgres. Is it still relevant or do you have a better or simpler way to do it ? Thanks

**Reply by r0m1CH** [reply]
I reply to myself. But yes this seems to work if you follow @elizhyu howto.
I need to check in the next coming days but everything seems to be okay.
Thanks !

**Reply by JC5** [reply]
Thank you, for sharing that this works :)

**Reply by r0m1CH** [reply]
One thing I noticed. I had to recreate a new token access for Firefly III Importer. The "old" but still valid token was there in my settings. However it was not working (status code is 401. Body is: {"message":"Unauthenticated.","exception":"AuthenticationException"}). Once recreate, no issue anymore.
